### ISSUE TITLE
User tables migrations

### ISSUE THREAD

**Issue by WoLfulus** [directus/api-next#137]
# Migrations

To enable migrations between different environments, we need to have a way to register all the changes the user makes to its tables. I don't think we should migrate `data` though, this is ONLY for structural changes, because data most of the time won't match between environments. If they match, it would be easier to just backup the database and restore in the target environment.

I'd suggest having something like the idea "event sourcing" for having that kind of stuff implemented, so we can replay the changes on another database.

Some points I'm concerned about:

### 1. Should we somehow have an integrity check to make sure we are running the migration over something we understand? 

Users might change tables outside Directus (which is where all the freedom of Directus comes from).

> I thought of having Directus running in two different modes, a `managed` and a `unmanaged` mode, in which in `managed`, Directus assumes all changes to tables are made over Directus api/app, while in `unmanaged`, Directus won't be managing any kind of migration between environments, and won't be able to do migrations.
> 
> Another option is having a `manual` migration step, in which changes to database can be manually registered in Directus so it can run for the user later. For example, user can add a `step` that executes an `UPDATE` query. Keep in mind that this is fragile, because most of the time it's not going to be cross-database compatible, and user won't be able to execute this migration if using `sqlite` for development and `postgres` for staging/production for example.

### 2. Directus upgrades should be treated as a migration `step`? 

There's always the case that the user made for example, 5 changes in the `development` environment, then upgraded Directus to another version, then made 7 more changes. I can see this breaking some installations depending on what changes from one version to another.

> I'm unsure what to think about this one. It's ok to not do this if you think about user data, as Directus isn't supposed to touch user tables, but I fear that metadata (for configs and stuff) can become out of sync or behave weirdly if this is not enforced.

### 3. How do we "connect" Directus instances so they know where/how to execute migrations?

> There are some ways that I could think of implementing this: one is having Directus projects to know each other, for for example `project-dev` can point to `project-staging`, and `project-staging` point to `project-prod`, and we can from one project, `push` the changes to another project, because now they know where to connect/execute the changes. We can also do the same thing, but as `pull` changes instead, in which the connections are made in reverse (`prod` knows about `staging`, and `staging` knows about `dev`) - I'd personally prefer this one because of the access levels.
> 
> Another way of doing this, is exporting all migration steps to a file, then `importing` to another project (this is a lot easier to implement and will also work for external/cloned projects too)

**Comment by rijkvanzanten** [comment]
> To enable migrations between different environments

Can you elaborate on this?

You're talking about only migrating the schema, not the data?

**Comment by WoLfulus** [comment]
Yes, in this case only the schema and database structure. Data doesn't make sense to be migrated (unless it's directus data like interface configs and etc)

**Comment by rijkvanzanten** [comment]
Wouldn't having just the before and after be enough in that case? That would allow us to calculate the diff and apply that to the old database to get to the new?

**Comment by WoLfulus** [comment]
Not really, because more stuff might happen in the middle of the process, and diffing databases can be error-prone. Some cases for example might need additional processing like setting a field to "required" which needs columns to have been updated first to get rid of any null values and such. Another scenario is the user renaming field `a` to `b`, then creating a new field `a`. Now all the values that was first at `a` would be empty on column `b`

**Comment by rijkvanzanten** [comment]
And this flow is to update one existing database to the state of another database, correct? Eg making sure production is up to date with staging

**Comment by WoLfulus** [comment]
Yes, we'd make changes on one stage/environment, then apply them to another stage/environment.

**Comment by rijkvanzanten** [comment]
I'm pretty sure we won't be able to accurately track all the changes if done directly through the database, so I think we'll have to slap a big fat warning on here 🤔 

Otherwise we'd have to do a snapshot of the "current state" at a certain interval so we'd be able to track manual changes?

**Comment by WoLfulus** [comment]
I don't think we'll ever be able to track manual changes at all, that's why I mentioned that we'd allow the user to insert changes manually as custom steps, or have a managed/unmanaged mode.

**Comment by rijkvanzanten** [comment]
I think a warning "Make sure you mirror all out-of-directus changes manually before proceeding" is enough for a MVP

**Comment by WoLfulus** [comment]
Yes, and as we're trying to cover too many use cases, both from experienced and non experienced users, having this managed/unmanaged makes sense because power users will have their own migrations/workflow anyways.

**Comment by carlobeltrame** [comment]
I'd like to propose a slightly different approach: Using the existing migration system from Laravel. It would work something like this:
When a developer changes something about the data structure through the Directus UI (e.g. sets a table to managed, configures an interface to manage a column, etc.), Directus generates a Laravel migration php file that describes that change. So in a way, I propose using Laravel migration files as a "log" or "event sourcing storage". When deploying Directus to another server, the developer would need to make sure that these generated migrations are also copied over and then applied via the Laravel migration tool (`artisan migrate`), in the correct order as dictated by the "timestamps" in the migration file names.

To address the questions above:
1. **Should we somehow have an integrity check to make sure we are running the migration over something we understand?**
I don't think we can do anything about the developer manually changing the database. But we can encourage the developer to only modify the database via his own Laravel migrations. If he does that, he gets the guarantee that Directus will continue working with his custom database as he intended in future Directus versions.
So in a way, my proposal is like your `manual` mode, except all the steps are described as Laravel migrations, so they are not fragile and are invariant to different DBMS (sqlite, postgres, etc.).

2. **Directus upgrades should be treated as a migration step?**
This could be solved by enforcing the correct ordering of migrations during a Directus upgrade, via the migration timestamps. So during a Directus upgrade, we'd need to change the timestamps of all new migrations from the upgrade to the current time (plus preserve their order of course).

3. **How do we "connect" Directus instances so they know where/how to execute migrations?**
It sounds very neat to be able to make some changes to the database structure locally via Directus, then deploy it to staging or production at the push of a button in the Directus UI. But **this would be very dangerous**, because it might break any API client running in staging or production. Directus is not a deployment tool, and shouldn't force a specific deployment workflow on the developer in my opinion. Deployments should run only when the code has been properly committed, unit tested, static analysed, manually tested, approved, grouped, tagged, or whatever else the developer's workflow might include. Saving any critical changes to disk for version control is in my eyes the easiest and best supported option for deploying a project to another environment.

4. **More stuff might happen during a migration than is evident from an SQL diff**
When setting a column to required (NOT NULL) via the Directus UI, we need enforce setting a default / fallback value, and that default should be reflected in the generated migration file via an UPDATE statement.
When the developer sets a column to required using his own custom Laravel migration, he is responsible for filling in the empty values of the column in his migration.

5. **Both experienced and non-experienced users should be supported**
For non-experienced users it is easy to use the Directus UI to get their DB to the state they need, and they can still deploy via SQL dump if they don't have SSH access to their shared hosting.
Experienced users can create their DB on their own using Laravel migrations if they so desire. Admittedly, this is effectively a lock-in to the Laravel migration system, as long as one wants to use Directus. In my opinion that is a limitation worth making, since the Laravel migration system is tried and tested, in contrast to reinventing the wheel in a Directus-specific custom migration system.

**Comment by WoLfulus** [comment]
Hey! Thanks for joining the discussion!

I agree with your points and while the next version of Directus are making way more easy and flexible for developers to use, we still need to cover the general use case of Directus where most people don't touch code.

With Laravel migrations most of these issues are solved and we'll definitely support the proposed Directus migrations though in a different way/format.

I think the main "issue" is in cases where the end user is using Directus in a way they can't/won't deal with code (using directus as an app only for example) - these must cover a way to export the changes made to the database (using Directus) and then imported to another Directus installation/server to apply the same structural changes/steps. 

Dumps doesn't work in these cases because as mentioned, you'll probably have different data in different environments.

**Comment by carlobeltrame** [comment]
Thanks for the explanation, sounds good to me. So if I understand correctly, you _do_ see Directus a kind of migrations tool / database management and deployment tool, kind of like a glorified phpmyadmin. I hadn't thought about it that way. As long as there is a well thought-out system for keeping changes under version control, that's fine with me (even though I'd really prefer to have just one migrations format and one migrations tool).

**Comment by WoLfulus** [comment]
Right now it's not a migration tool, but we feel like it should support some kind of export/migration system because it's a frequent request on how to transition the changes between environments.

I experienced this myself, and transitioning the changes from my local development environment to the production server is always a pain (not the schema, but the Directus configs itself). 

We'll iterate a lot more on this topic soon.

**Comment by WoLfulus** [comment]
FWD from Discussions:

```
I haven't iterated on this feature just yet. but I was playing with something related last week.

While I still want to provide some kind of migration system within Directus itself, something like this (https://github.com/directus/api-next/blob/wolfulus/examples/blog.yml) can be used to reconfigure Directus metadata through `php artisan directus:import blog.yml`, while Laravel migrations can be used to migrate the database structure. 

I'm also planning on exposing some traits/methods from Directus itself to be written inside user's Laravel migrations (that might be the best solution for more advanced users, while an import/export might work just fine for non-tech people)
```
